x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java

```diff
@@ -34,40 +37,80 @@
     public static final String CONDITIONAL_SKIP_SHRINK_STEP = BranchingStep.NAME + "-check-prerequisites";
     public static final String CONDITIONAL_DATASTREAM_CHECK_KEY = BranchingStep.NAME + "-on-datastream-check";
 
     private static final ConstructingObjectParser<ShrinkAction, Void> PARSER =
-        new ConstructingObjectParser<>(NAME, a -> new ShrinkAction((Integer) a[0]));
+        new ConstructingObjectParser<>(NAME, a -> new ShrinkAction((Integer) a[0], (ByteSizeValue) a[1]));
 
     static {
-        PARSER.declareInt(ConstructingObjectParser.constructorArg(), NUMBER_OF_SHARDS_FIELD);
+        PARSER.declareInt(ConstructingObjectParser.optionalConstructorArg(), NUMBER_OF_SHARDS_FIELD);
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> ByteSizeValue.parseBytesSizeValue(p.text(), MAX_SINGLE_PRIMARY_SHARD_SIZE.getPreferredName()),
+            MAX_SINGLE_PRIMARY_SHARD_SIZE, ObjectParser.ValueType.STRING);
     }
 
-    private int numberOfShards;
+    private Integer numberOfShards;
+    private ByteSizeValue maxSinglePrimaryShardSize;
 
     public static ShrinkAction parse(XContentParser parser) throws IOException {
         return PARSER.parse(parser, null);
     }
 
-    public ShrinkAction(int numberOfShards) {
-        if (numberOfShards <= 0) {
-            throw new IllegalArgumentException("[" + NUMBER_OF_SHARDS_FIELD.getPreferredName() + "] must be greater than 0");
+    public ShrinkAction(@Nullable Integer numberOfShards, @Nullable ByteSizeValue maxSinglePrimaryShardSize) {
+        if (numberOfShards != null && maxSinglePrimaryShardSize != null) {
+            throw new IllegalArgumentException("Cannot set both [number_of_shards] and [max_single_primary_shard_size]");
+        }
+        if (numberOfShards == null && maxSinglePrimaryShardSize == null) {
+            throw new IllegalArgumentException("Either [number_of_shards] or [max_single_primary_shard_size] must be set");
+        }
+        if (maxSinglePrimaryShardSize != null) {
+            if (maxSinglePrimaryShardSize.getBytes() <= 0) {
+                throw new IllegalArgumentException("[max_single_primary_shard_size] must be greater than 0");
+            }
+            this.maxSinglePrimaryShardSize = maxSinglePrimaryShardSize;
+        } else {
+            if (numberOfShards <= 0) {
+                throw new IllegalArgumentException("[" + NUMBER_OF_SHARDS_FIELD.getPreferredName() + "] must be greater than 0");
+            }
+            this.numberOfShards = numberOfShards;
         }
-        this.numberOfShards = numberOfShards;
     }
 
     public ShrinkAction(StreamInput in) throws IOException {
-        this.numberOfShards = in.readVInt();
+        if (in.readBoolean()) {
+            this.numberOfShards = in.readVInt();
+        } else {
+            this.numberOfShards = null;
+        }
+        if (in.readBoolean()) {
```

Comment:
no need for the second boolean, since the values are mutually exclusive (my above example removes this)